### PR TITLE
Merge branch 'main' of https://github.com/asgatech20/AndroidDateFormatter into development

### PR DESCRIPTION
**Pull request by Ahmed-Saber-25**

### CODE REVIEW

File: dateformatterutil/src/main/java/com/example/dateformatter/DateFormatterUtil.kt [hunk 1]
```diff
@@ -0,0 +1,182 @@
...
+
+
+object DateFormatterUtil {
+    //enum dateformat
+    /**
+     * @usage this function is used to convert date of type string to date based on required date pattern
+     * @param from A String containing the date string
+     * @param outputDateFormatPattern A String represents required date format
+     * @return A Date value that represents the output date
+     */
```
Comment (asgatech20) [inline dateformatterutil/src/main/java/com/example/dateformatter/DateFormatterUtil.kt]: Fix the documentation text as doesn't describe the function usage correctly.

File: dateformatterutil/src/main/java/com/example/dateformatter/DateFormatterUtil.kt [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+            print(ex.printStackTrace())
+        }
+        return null
+
+    }
+
+    /**
+     * @usage this function is used to convert date of type Date to date String based on required date pattern
+     * @param from A Date containing the date object
+     * @param format A String represents input date format
```
Comment (asgatech20) [inline dateformatterutil/src/main/java/com/example/dateformatter/DateFormatterUtil.kt]: Change to be '* @param format A String represents the required format'

File: dateformatterutil/src/main/java/com/example/dateformatter/DateFormatterUtil.kt [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+import org.joda.time.chrono.GregorianChronology
+import org.joda.time.chrono.ISOChronology
+import org.joda.time.chrono.IslamicChronology
+import java.text.ParseException
+import java.text.SimpleDateFormat
+import java.util.*
+
+
+object DateFormatterUtil {
+    //enum dateformat
```
Comment (asgatech20) [inline dateformatterutil/src/main/java/com/example/dateformatter/DateFormatterUtil.kt]: Remove this comment and write usage documentation for the object

File: dateformatterutil/src/main/java/com/example/dateformatter/DateFormatterUtil.kt [hunk 4]
```diff
@@ -0,0 +1,182 @@
...
+
+    /**
+     * @usage this function is used to convert date of type Date to hijri date String
+     * @param from A Date containing the date object
+     * @return A LocalDate object that represents the hijri date
+     */
+    fun convertToHijriDate(from: Date): LocalDate? {
+        try {
+            val calendar = Calendar.getInstance()
+            calendar.timeInMillis = from.time
```
Comment (asgatech20) [inline dateformatterutil/src/main/java/com/example/dateformatter/DateFormatterUtil.kt]: Convert to separate function

File: dateformatterutil/src/main/java/com/example/dateformatter/DateFormatterUtil.kt [hunk 5]
```diff
@@ -0,0 +1,182 @@
...
+            val iso: Chronology = GregorianChronology.getInstance()
+            val hijri: Chronology = IslamicChronology.getInstance()
+
+            val todayIso = LocalDate(
+                calendar[Calendar.YEAR],
+                calendar[Calendar.MONTH] + 1,
+                calendar[Calendar.DAY_OF_MONTH], iso
+            )
```
Comment (asgatech20) [inline dateformatterutil/src/main/java/com/example/dateformatter/DateFormatterUtil.kt]: Concert to separate function

File: dateformatterutil/src/main/java/com/example/dateformatter/DateFormatterUtil.kt [hunk 6]
```diff
@@ -0,0 +1,182 @@
...
+     * @usage this function is used to convert LocalDateTime object to hijri date String
+     * @param localDate A LocalDate containing the date object
+     * @param isIslamic A Boolean confirm that the date object
+     * @return A Date that represents the localDate is islamic or normal date
+     */
+    fun convertFromIslamic(localDate: LocalDate, isIslamic: Boolean): Date? {
+        return if (isIslamic) {
+            val iso: Chronology = ISOChronology.getInstanceUTC()
+            val dtIso = localDate.toDateTimeAtStartOfDay(DateTimeZone.UTC)
+                .withChronology(iso).toLocalDate()
```
Comment (asgatech20) [inline dateformatterutil/src/main/java/com/example/dateformatter/DateFormatterUtil.kt]: Convert to separate function with return type

File: dateformatterutil/src/main/java/com/example/dateformatter/DateFormatterUtil.kt [hunk 7]
```diff
@@ -0,0 +1,182 @@
...
+     * @usage this function is used to convert LocalDateTime object to hijri date String
+     * @param from A LocalDateTime containing the datetime object
+     * @return A String that represents the date
+     */
+    fun convert(date: LocalDateTime, format: String): String {
+        return date.toString(format)
+    }
+
+    /**
+     * @usage this function is used to convert LocalDateTime object to hijri date String
```
Comment (asgatech20) [inline dateformatterutil/src/main/java/com/example/dateformatter/DateFormatterUtil.kt]: change LocalDateTime  to the correct type that function already accepts

File: dateformatterutil/src/main/java/com/example/dateformatter/DateFormatterUtil.kt [hunk 8]
```diff
@@ -0,0 +1,182 @@
...
+        format.timeZone = TimeZone.getTimeZone("UTC")
+        val sdf = SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss")
+        var d: Date? = null
+        try {
+            d = sdf.parse(dateString)
+        } catch (e: ParseException) {
+            e.printStackTrace()
+        }
+        return format.format(d)
+    }
```
Comment (asgatech20) [inline dateformatterutil/src/main/java/com/example/dateformatter/DateFormatterUtil.kt]: Refactor this function to be more usable as from the function body it is seen that user can't figure out he must pass a dateString with specific format that the function accepts "yyyy-MM-dd'T'HH:mm:ss"

File: dateformatterutil/src/main/java/com/example/dateformatter/DateFormatterUtil.kt [hunk 9]
```diff
@@ -0,0 +1,182 @@
...
+        format.timeZone = TimeZone.getTimeZone("UTC")
+        val sdf = SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss")
+        var d: Date? = null
+        try {
+            d = sdf.parse(date)
+        } catch (e: ParseException) {
+            e.printStackTrace()
+        }
+        return format.format(d)
+    }
```
Comment (asgatech20) [inline dateformatterutil/src/main/java/com/example/dateformatter/DateFormatterUtil.kt]: Refactor this function to be more usable as from the function body it is seen that user can't figure out he must pass a dateString with specific format that the function accepts "yyyy-MM-dd'T'HH:mm:ss"

File: dateformatterutil/src/main/java/com/example/dateformatter/DateFormatterUtil.kt [hunk 10]
```diff
@@ -0,0 +1,182 @@
...
+    /**
+     * @usage this function is used to compare two date
+     * @param fromDate A String containing the start date
+     * @param toDate A String containing the end date
+     * @return A Boolean that indicates the end date is after start or not
+     */
+    fun compareDates(fromDate: String , toDate: String): Boolean {
```
Comment (asgatech20) [inline dateformatterutil/src/main/java/com/example/dateformatter/DateFormatterUtil.kt]: Use an appropriate function naming as the current function name doesn't represent the actual usage of the function.

File: dateformatterutil/src/main/java/com/example/dateformatter/DateFormatterUtil.kt [hunk 11]
```diff
@@ -0,0 +1,182 @@
...
+     * @usage this function is used to split hours and min from date based on am and pm
+     * @param date A String containing the date
+     * @return A String that contains hh:min am/pm
+     */
+    fun getDatFromPmAmHoursMin(date: String): String {
+        val h1 = date.split(":".toRegex()).toTypedArray()
+        val hour = h1[0]
+        val minute = h1[1]
+        return if (date.contains("PM")) "$hour:$minute PM" else "$hour:$minute AM"
+    }
```
Comment (asgatech20) [inline dateformatterutil/src/main/java/com/example/dateformatter/DateFormatterUtil.kt]: First: Use an appropriate function naming as the current function name doesn't represent the actual usage of the function.
Second: This function is not null safe (if user passes a dateString that doesn't have hours and minutes format, the function will fail and application will crash), so it needs to be refactored.

File: dateformatterutil/src/main/java/com/example/dateformatter/DateFormatterUtil.kt [hunk 12]
```diff
@@ -0,0 +1,182 @@
...
+    /**
+     * @usage this function is used to get current date
+     * @return A String that represents the current date
+     */
+    fun getCurrentData(): String {
+        val c = Calendar.getInstance().time
+        val df = SimpleDateFormat("yyyy/MM/dd", Locale.ENGLISH)
+        return df.format(c)
+    }
```
Comment (asgatech20) [inline dateformatterutil/src/main/java/com/example/dateformatter/DateFormatterUtil.kt]: Add a param that makes user capable to get the current date with a certain format.

File: dateformatterutil/src/main/java/com/example/dateformatter/DateFormatterUtil.kt [hunk 13]
```diff
@@ -0,0 +1,182 @@
...
+                SimpleDateFormat("yyyy/MM/dd - hh:mm:ss a", Locale.ENGLISH)
+            val dateLong = df.parse(georgianDate).time
+            val formatter = SimpleDateFormat("hh:mm a", Locale.ENGLISH)
+            calendar.timeInMillis = dateLong
+            return formatter.format(calendar.time)
+        } catch (e: ParseException) {
+            e.printStackTrace()
+        }
+        return ""
+    }
```
Comment (asgatech20) [inline dateformatterutil/src/main/java/com/example/dateformatter/DateFormatterUtil.kt]: First: Use an appropriate function naming as the current function name doesn't represent the actual usage of the function.
Second: The user can't figure out he must pass a dateString with specific format that the function accepts "yyyy/MM/dd - hh:mm:ss a"
=> for example: what should happen if he passes a dateString with format of  "yyyy-MM-dd'T'HH:mm:ss" ?

File: dateformatterutil/src/test/java/com/example/dateformatter/DateFormatterUtilTest.kt [hunk 14]
```diff
@@ -0,0 +1,114 @@
...
+    fun convertToHijriDate() {
+        assertEquals(
+            com.example.dateformatter.DateFormatterUtil.convertToHijriDate(
+                com.example.dateformatter.DateFormatterUtil.convertStringToDate(
+                    "12.01.2022 01:32:27",
+                    "dd.MM.yyyy HH:mm:ss"
+                )!!
+            )?.year, LocalDate("1443-06-08").year
+        )
+    }
```
Comment (asgatech20) [inline dateformatterutil/src/test/java/com/example/dateformatter/DateFormatterUtilTest.kt]: The data of 12 / 01 /2021 should return 9 / 6 / 1443 in Higri not 8 / 6 / 1443 
Fix the function in your formatter class

File: dateformatterutil/src/test/java/com/example/dateformatter/DateFormatterUtilTest.kt [hunk 15]
```diff
@@ -0,0 +1,114 @@
...
+import org.joda.time.LocalDate
+import org.joda.time.LocalDateTime
+import org.joda.time.tz.UTCProvider
+import org.junit.Assert.*
+import org.junit.Before
+import org.junit.Test
+import java.util.*
+
+
+class DateFormatterUtilTest{
```
Comment (asgatech20) [inline dateformatterutil/src/test/java/com/example/dateformatter/DateFormatterUtilTest.kt]: General Comment:
Add Test functions for the non acceptable cases.

**Review by asgatech20 — CHANGES_REQUESTED** [review]
- Please Add Readme file that should contain docs of the component.
- Provide an example of using the component in the example app.
- Take care of functions naming.
- Add Test functions for the non acceptable cases in the DateFormatterUtilTest class.

File: dateformatterutil/src/main/java/com/example/dateformatter/DateFormatterUtil.kt (hunk 1 above)
Comment (Ahmed-Saber-25) [reply dateformatterutil/src/main/java/com/example/dateformatter/DateFormatterUtil.kt]: resolved

**Comment by Ahmed-Saber-25** [comment]
solve commits

File: dateformatterutil/src/main/java/com/example/dateformatter/DateFormatterUtil.kt [hunk 16]
```diff
@@ -107,17 +107,11 @@ object DateFormatterUtil {
      * @usage this function is used to convert LocalDate object to hijri date String
      * @acceptedFormats all formats in StandardDateParser enum class if you enter invalid value for entered pattern it will throw exception
      * @param localDate A LocalDate containing the date object
-     * @param isIslamic A Boolean confirm that the date object
      * @return A Date that represents the localDate is islamic or normal date
```
Comment (asgatech20) [inline dateformatterutil/src/main/java/com/example/dateformatter/DateFormatterUtil.kt]: correct the documentation text

File: dateformatterutil/src/main/java/com/example/dateformatter/DateFormatterUtil.kt [hunk 17]
```diff
@@ -131,47 +125,6 @@ object DateFormatterUtil {
             .withChronology(iso).toLocalDate()
```
Comment (asgatech20) [inline dateformatterutil/src/main/java/com/example/dateformatter/DateFormatterUtil.kt:140]: convert to minusDays(1)

**Review by asgatech20 — DISMISSED** [review]
- Try to use exceptions to handle critical failures.
- Write test cases for the wrong scenarios.

File: README.md [hunk 18]
```diff
@@ -1 +1,182 @@
-# AndroidDateFormatter
\ No newline at end of file
+# AndroidDateFormatter
+     this module is used to convert date object to string value with specific formate ,hijri-to-gregorian and vice versa and extract any part of date you need
```
Comment (IbrahimAli2017) [inline README.md]: can convert from string to date or not?

File: README.md [hunk 19]
```diff
@@ -1 +1,182 @@
...
+     1- start of calender : hijri-to-gregorian: 1389-10-23=>1970-01-01
+     2- all formats in StandardDateParser enum class if you enter invalid value for entered pattern it will throw exception
+## Usage
+
+### fun convertStringToDate(from: String, dateParser: StandardDateParser): Date?
+    this function is used to parse string date to Date Object based on passed date format
+ 
+**Params**
+* `from` **{String}** mandatory input
+* `dateParser` **{StandardDateParser}** mandatory input
```
Comment (IbrahimAli2017) [inline README.md:15]: what is dateParser?
should define the DateParser before starting describe  the functions

File: README.md [hunk 20]
```diff
@@ -1 +1,182 @@
...
+* `returns` **{Date}**
+
+**Example**
+
```
Comment (IbrahimAli2017) [inline README.md]: add description about the example

File: README.md [hunk 21]
```diff
@@ -1 +1,182 @@
...
+### fun convertDateToString(from: Date?, dateParser: StandardDateParser): String
+    this function is used to format date of type Date to date String based on required date format
+ 
+**Params**
+* `from` **{Date}** mandatory input
+* `dateParser` **{StandardDateParser}** mandatory input
+* `returns` **{String}**
+
+**Example**
+
```
Comment (IbrahimAli2017) [inline README.md:39]: add description about example

File: README.md [hunk 22]
```diff
@@ -1 +1,182 @@
...
+```kt
+val date = DateFormatterUtil.convertStringToDate(
```
Comment (IbrahimAli2017) [inline README.md]: please build Date Object using calendar not from convertStringToDate
this example to show how convert date to string not string to another string

File: README.md [hunk 23]
```diff
@@ -1 +1,182 @@
...
+
+### fun convertToHijriDate(from: Date): LocalDate? 
+    this function is used to convert date of type Date to hijri date String
+ 
+**Params**
+* `from` **{Date}** mandatory input
+* `returns` **{LocalDate}**
+
+**Example**
+
```
Comment (IbrahimAli2017) [inline README.md:60]: add description about the example

File: README.md [hunk 24]
```diff
@@ -1 +1,182 @@
...
+### fun convert(date: LocalDateTime, dateParser: StandardDateParser): String 
+    this function is used to convert LocalDateTime object to date String
+ 
+**Params**
+* `from` **{LocalDateTime}** mandatory input
+* `dateParser` **{StandardDateParser}** mandatory input
+* `returns` **{String}**
+
+**Example**
+
```
Comment (IbrahimAli2017) [inline README.md:84]: the same comment

File: README.md [hunk 25]
```diff
@@ -1 +1,182 @@
...
+### convertFromIslamic(localDate: LocalDate): Date? 
+    this function is used to convert LocalDate object to hijri date String
+ 
+**Params**
+* `localDate` **{LocalDate}** mandatory input
+* `returns` **{Date?}**
+
+**Example**
+
```
Comment (IbrahimAli2017) [inline README.md:110]: same comment

File: README.md [hunk 26]
```diff
@@ -1 +1,182 @@
...
+            StandardDateParser.DD_MM_YYYY
+        )
+        val hijri = DateFormatterUtil.convertToHijriDate(date!!)
+        DateFormatterUtil.convertFromIslamic(hijri!!)
+      
+      
+//=>  Wed Jan 12 00:00:00 EET 2022 
+```
+
+### fun isFutureDate(fromDate: String, toDate: String): Boolean
```
Comment (IbrahimAli2017) [inline README.md]: rename to be is after
change fromDate to be firstDate
change toDate to be secondDate

File: README.md [hunk 27]
```diff
@@ -1 +1,182 @@
...
+* `returns` **{String}**
+
+**Example**
+
+```kt
+    DateFormatterUtil.getCurrentData(StandardDateParser.YYYY_MM_DD)
+    //=>2022-01-24
+
+```
+
```
Comment (IbrahimAli2017) [inline README.md:187]: add contributors

File: dateformatterutil/src/main/java/com/example/dateformatter/DateFormatterUtil.kt [hunk 28]
```diff
@@ -0,0 +1,210 @@
...
+            val hijri: Chronology = IslamicChronology.getInstance()
+            val todayIso = initTodayIso(calendar)
+            val todayHijri = LocalDate(
+                todayIso.toDateTimeAtStartOfDay(),
+                hijri
+            )
+            return todayHijri
+        } catch (ex: ParseException) {
+            ExceptionLogger.Error(ex.localizedMessage,ex)
+            ""
```
Comment (IbrahimAli2017) [inline dateformatterutil/src/main/java/com/example/dateformatter/DateFormatterUtil.kt]: this for what?

File: dateformatterutil/src/main/java/com/example/dateformatter/DateFormatterUtil.kt [hunk 29]
```diff
@@ -0,0 +1,210 @@
...
+        return ""
+    }
+
+    /**
+     * @usage this function is used to get current date
+     * @acceptedFormats all formats in StandardDateParser enum class
+     * @param dateParser A StandardDateParser enum class contain all valid date and time format
+     * @return A String that represents the current date
+     */
+    fun getCurrentData(dateParser: StandardDateParser): String {
```
Comment (IbrahimAli2017) [inline dateformatterutil/src/main/java/com/example/dateformatter/DateFormatterUtil.kt]: rename the function to be getCurrentDate not Data

File: dateformatterutil/src/main/java/com/example/dateformatter/DateFormatterUtil.kt [hunk 30]
```diff
@@ -0,0 +1,210 @@
...
+     * @acceptedFormat HH:MM:ss AM/PM
+     * @param time A String containing the date
+     * @return A String that contains hh:min am/pm
+     */
+    fun getHoursMinFromTime(time: String): String {
+        try {
+            val h1 = time.split(":".toRegex()).toTypedArray()
+            val hour = h1[0]
+            val minute = h1[1]
+            return if (hour.toInt() in 13..23 || time.contains("PM")) "$hour:$minute PM" else "$hour:$minute AM"
```
Comment (IbrahimAli2017) [inline dateformatterutil/src/main/java/com/example/dateformatter/DateFormatterUtil.kt]: what if the string in ar lang the contains function will be false 
PM and AM good for en Lang we should support AR Lang

File: dateformatterutil/src/main/java/com/example/dateformatter/DateFormatterUtil.kt [hunk 31]
```diff
@@ -0,0 +1,210 @@
...
+
+    /**
+     * @usage this function is used to check to-date is after from-date
+     * @acceptedFormats all date or date-time formats in StandardDateParser enum class if you enter invalid value for entered pattern it will
+     * throw exception or if you pass time only it will throw IllegalArgumentException
+     * @param fromDate A String containing the start date
+     * @param toDate A String containing the end date
+     * @return A Boolean that indicates the end date is after start or not
+     */
+    fun isFutureDate(fromDate: String, toDate: String): Boolean {
```
Comment (IbrahimAli2017) [inline dateformatterutil/src/main/java/com/example/dateformatter/DateFormatterUtil.kt]: rename the function
isFutureDate => isAfter
fromDate => firstDate
toDate => secondDate

File: dateformatterutil/src/main/java/com/example/dateformatter/DateFormatterUtil.kt [hunk 32]
```diff
@@ -0,0 +1,210 @@
...
+        return try {
+            Date(toDate).after(Date(fromDate))
```
Comment (IbrahimAli2017) [inline dateformatterutil/src/main/java/com/example/dateformatter/DateFormatterUtil.kt]: what is the deference between 
Date(toDate) & dateParser.parser.parse(from)

File: dateformatterutil/src/test/java/com/example/dateformatter/DateFormatterUtilTest.kt [hunk 33]
```diff
@@ -0,0 +1,269 @@
...
+            "22:17 PM",
+            DateFormatterUtil.getHoursMinFromTime("22:17")
+        )
+        assertEquals(
+            "02:17 AM",
+            DateFormatterUtil.getHoursMinFromTime("02:17:00 AM")
+        )
+    }
+    @Test
+    fun getDateFromPmAmHoursMin_return_invalid_result_when_passing_invalid_params() {
```
Comment (IbrahimAli2017) [inline dateformatterutil/src/test/java/com/example/dateformatter/DateFormatterUtilTest.kt]: you should test AR Lang

File: dateformatterutil/src/test/java/com/example/dateformatter/DateFormatterUtilTest.kt [hunk 34]
```diff
@@ -0,0 +1,269 @@
...
+        )
+
+         //java.lang.IllegalArgumentException
+//        assertEquals(
+//            true,
+//            DateFormatterUtil.isFutureDate("12:08", "01:12")
+//        )
+    }
+    @Test
+    fun getDateFromPmAmHoursMin_return_valid_result_when_passing_valid_params() {
```
Comment (IbrahimAli2017) [inline dateformatterutil/src/test/java/com/example/dateformatter/DateFormatterUtilTest.kt]: should test AR Lang

File: dateformatterutil/src/main/java/com/example/dateformatter/DateFormatterUtil.kt (hunk 32 above)
Comment (Ahmed-Saber-25) [reply dateformatterutil/src/main/java/com/example/dateformatter/DateFormatterUtil.kt]: first : here we restricted to date string and it is deprecated
 @Deprecated
    public Date(String s) {
        this(parse(s));
    }

second : here we can pass format and local 
SimpleDateFormat("dd.MM.yyyy HH:mm:ss",Locale.ENGLISH).parse(dateString)

File: dateformatterutil/src/main/java/com/example/dateformatter/DateFormatterUtil.kt [hunk 35]
```diff
@@ -0,0 +1,193 @@
...
+
+    /**
+     * @usage this function is used to check to-date is after from-date
+     * @acceptedFormats all date or date-time formats in StandardDateParser enum class if you enter invalid value for entered pattern it will
+     * throw exception or if you pass time only it will throw IllegalArgumentException
+     * @param firstDate A String containing the start date
+     * @param secondDate A String containing the end date
+     * @return A Boolean that indicates the end date is after start or not
+     */
+    fun isAfterDate(firstDate: String, secondDate: String): Boolean {
```
Comment (IbrahimAli2017) [inline dateformatterutil/src/main/java/com/example/dateformatter/DateFormatterUtil.kt]: it should use the previous functions to get date

**Review by IbrahimAli2017 — APPROVED** [review]

**Pull request merged**